### PR TITLE
Add support for maps

### PR DESCRIPTION
**Pull request by johnstairs**
Adding a map/dictionary datatype. 

The shorthand syntax is
```
string->int
```
And the full syntax (which allows more complicated value types):
```yaml
!map
keys: string
values: int
```

They map to an `std::unordered_map<K, V>`, support generic type parameters, and can also be used in computed expressions (`size(myMap)` or `myMap["myKey"]`).

Addresses #47

### CODE REVIEW

File: cpp/test/computed_fields_test.cc [hunk 1]
```diff
@@ -35,6 +35,12 @@ TEST(ComputedFieldsTests, AccessFields) {
 
   r.vector_of_vectors_field = {{1, 2, 3}, {4, 5, 6}};
   EXPECT_EQ(r.AccessVectorOfVectorsField(), r.vector_of_vectors_field[1][2]);
+
+  r.map_field = {{"hello", "world"}, {"world", "bye"}};
+  EXPECT_EQ(r.AccessMap(), r.map_field);
+  EXPECT_EQ(r.AccessMapEntry(), "world");
+  EXPECT_EQ(r.AccessMapEntryWithComputedField(), "world");
```
Comment (hansenms) [inline cpp/test/computed_fields_test.cc:42]: I am wondering if we should have a test that confirms that we throw if the key is not there. Right now we rely on https://en.cppreference.com/w/cpp/container/unordered_map/at to do that, so we are pretty sure that is the case, but if we later on switch the underlying implementation or we use this set of tests as a template later to write tests in another language, then we might miss it. Note that the `[]` operator of unordered_map creates the entry if it doesn't exist (https://en.cppreference.com/w/cpp/container/unordered_map/operator_at).

File: tooling/pkg/dsl/yaml.go [hunk 2]
```diff
@@ -659,6 +659,46 @@ func UnmarshalStreamYAML(value *yaml.Node) (*GeneralizedType, error) {
 	return t, nil
 }
 
+func UnmarshalMapYAML(value *yaml.Node) (*GeneralizedType, error) {
+	if value.Kind != yaml.MappingNode {
+		return nil, parseError(value, "a !map must be specified with fields `keys` and optionally `values`")
```
Comment (hansenms) [inline tooling/pkg/dsl/yaml.go:664]: I think both `keys` and `values` are mandatory fields. We did discuss keys defaulting to string, but it doesn't currently right? I think that is the way it should be.

**Review by hansenms — APPROVED** [review]
This looks great. Really nice with the short-hand for these too.

File: tooling/pkg/dsl/yaml.go (hunk 2 above)
Comment (johnstairs) [reply tooling/pkg/dsl/yaml.go:664]: Ah yes, that error message is wrong. Will fix.

**Pull request merged**